### ISSUE TITLE
Item not found / Failed to update last connected value for device

### ISSUE THREAD

**Issue by ajcasagrande** [edgexfoundry/edgex-go#3635]
# 🐞 Bug Report

### Affected Services

- Metadata
- Device-SDK-Go

### Is this a regression?

Not sure

#### Combined Logs
<pre><code>
30511\",\"sender\":\"edgex-core-metadata\",\"category\":\"SW_HEALTH\",\"severity\":\"NORMAL\",\"content\":\"Device update: SpeedwayR-11-43-62-POST\",\"description\":\"Metadata device notice\",\"labels\":[\"metadata\"]}"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.536641184Z app=edgex-device-rfid-llrp source=device.go:72 msg="Added device: SpeedwayR-11-43-62"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.536763745Z app=edgex-device-rfid-llrp source=driver.go:717 device=SpeedwayR-11-43-62 msg="Creating new connection for device."
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.53991939Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgReaderEventNotification device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
edgex-device-rfid-llrp | level=WARN ts=2021-07-22T15:27:10.540009732Z app=edgex-device-rfid-llrp source=logging.go:38 message-version=Version1_0_1 client-version=Version1_1 msg="LLRP incoming message version mismatch"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.540245141Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgGetSupportedVersion device=SpeedwayR-11-43-62 msg="Sending LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.541450023Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgErrorMessage device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.541618434Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgSetReaderConfig device=SpeedwayR-11-43-62 msg="Sending LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.545164887Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgSetReaderConfigResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
edgex-app-service-rfid-llrp-inventory | level=INFO ts=2021-07-22T15:27:10.548700796Z app=rfid-llrp-inventory source=controller.go:123 msg="Adding device to default group: SpeedwayR-11-43-62"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.550290048Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgGetReaderCapabilities device=SpeedwayR-11-43-62 msg="Sending LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.553602604Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgGetReaderCapabilitiesResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.55681481Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgCustomMessage device=SpeedwayR-11-43-62 msg="Sending LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.558545516Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgCustomMessage device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
edgex-core-metadata    | level=ERROR ts=2021-07-22T15:27:10.560519413Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.560269701Z app=edgex-device-rfid-llrp source=device.go:229 forcedKA=30s msg="Adding device-service-enforced a KeepAlive spec to ReaderConfig."
edgex-device-rfid-llrp | level=ERROR ts=2021-07-22T15:27:10.560722887Z app=edgex-device-rfid-llrp source=utils.go:302 msg="Failed to update last connected value for device: SpeedwayR-11-43-62"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.560648883Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgSetReaderConfig device=SpeedwayR-11-43-62 msg="Sending LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.613430752Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgSetReaderConfigResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.617132179Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgDeleteROSpec device=SpeedwayR-11-43-62 msg="Sending LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.620035882Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgDeleteROSpecResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.625388649Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgAddROSpec device=SpeedwayR-11-43-62 msg="Sending LLRP message"
edgex-device-rfid-llrp | level=INFO ts=2021-07-22T15:27:10.632810102Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgAddROSpecResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
edgex-core-metadata    | level=ERROR ts=2021-07-22T15:27:10.650220089Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
edgex-device-rfid-llrp | level=ERROR ts=2021-07-22T15:27:10.653529792Z app=edgex-device-rfid-llrp source=utils.go:302 msg="Failed to update last connected value for device: SpeedwayR-11-43-62"
</code></pre>

#### device-rfid-llrp logs
<pre><code>
level=INFO ts=2021-07-22T15:27:08.516307224Z app=edgex-device-rfid-llrp source=discover.go:396 host=192.168.0.105 port=5084 msg="Connection initiated successfully."
level=INFO ts=2021-07-22T15:27:08.516391982Z app=edgex-device-rfid-llrp source=discover.go:430 msg="Discovered device: &{deviceName:SpeedwayR-11-43-62 host:192.168.0.105 port:5084 vendor:25882 model:2001002 fwVersion:6.4.1.240}"
level=INFO ts=2021-07-22T15:27:08.516576255Z app=edgex-device-rfid-llrp source=manageddevices.go:71 msg="Device SpeedwayR-11-43-62 cannot be found in cache"
level=INFO ts=2021-07-22T15:27:10.499203383Z app=edgex-device-rfid-llrp source=driver.go:892 msg="Discovered 1 new devices in 2.086348919s."
level=INFO ts=2021-07-22T15:27:10.499355097Z app=edgex-device-rfid-llrp source=async.go:140 msg="Adding discovered device SpeedwayR-11-43-62 to Edgex"
level=INFO ts=2021-07-22T15:27:10.536641184Z app=edgex-device-rfid-llrp source=device.go:72 msg="Added device: SpeedwayR-11-43-62"
level=INFO ts=2021-07-22T15:27:10.536763745Z app=edgex-device-rfid-llrp source=driver.go:717 device=SpeedwayR-11-43-62 msg="Creating new connection for device."
level=INFO ts=2021-07-22T15:27:10.53991939Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgReaderEventNotification device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
level=WARN ts=2021-07-22T15:27:10.540009732Z app=edgex-device-rfid-llrp source=logging.go:38 message-version=Version1_0_1 client-version=Version1_1 msg="LLRP incoming message version mismatch"
level=INFO ts=2021-07-22T15:27:10.540245141Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgGetSupportedVersion device=SpeedwayR-11-43-62 msg="Sending LLRP message"
level=INFO ts=2021-07-22T15:27:10.541450023Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgErrorMessage device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
level=INFO ts=2021-07-22T15:27:10.541618434Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgSetReaderConfig device=SpeedwayR-11-43-62 msg="Sending LLRP message"
level=INFO ts=2021-07-22T15:27:10.545164887Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgSetReaderConfigResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
level=INFO ts=2021-07-22T15:27:10.550290048Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgGetReaderCapabilities device=SpeedwayR-11-43-62 msg="Sending LLRP message"
level=INFO ts=2021-07-22T15:27:10.553602604Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgGetReaderCapabilitiesResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
level=INFO ts=2021-07-22T15:27:10.55681481Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgCustomMessage device=SpeedwayR-11-43-62 msg="Sending LLRP message"
level=INFO ts=2021-07-22T15:27:10.558545516Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgCustomMessage device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
level=INFO ts=2021-07-22T15:27:10.560269701Z app=edgex-device-rfid-llrp source=device.go:229 forcedKA=30s msg="Adding device-service-enforced a KeepAlive spec to ReaderConfig."
level=ERROR ts=2021-07-22T15:27:10.560722887Z app=edgex-device-rfid-llrp source=utils.go:302 msg="Failed to update last connected value for device: SpeedwayR-11-43-62"
level=INFO ts=2021-07-22T15:27:10.560648883Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgSetReaderConfig device=SpeedwayR-11-43-62 msg="Sending LLRP message"
level=INFO ts=2021-07-22T15:27:10.613430752Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgSetReaderConfigResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
level=INFO ts=2021-07-22T15:27:10.617132179Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgDeleteROSpec device=SpeedwayR-11-43-62 msg="Sending LLRP message"
level=INFO ts=2021-07-22T15:27:10.620035882Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgDeleteROSpecResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
level=INFO ts=2021-07-22T15:27:10.625388649Z app=edgex-device-rfid-llrp source=logging.go:25 type=MsgAddROSpec device=SpeedwayR-11-43-62 msg="Sending LLRP message"
level=INFO ts=2021-07-22T15:27:10.632810102Z app=edgex-device-rfid-llrp source=logging.go:34 type=MsgAddROSpecResponse device=SpeedwayR-11-43-62 msg="Incoming LLRP message"
level=ERROR ts=2021-07-22T15:27:10.653529792Z app=edgex-device-rfid-llrp source=utils.go:302 msg="Failed to update last connected value for device: SpeedwayR-11-43-62"
</code></pre>

#### edgex-core-metadata logs
<pre><code>
level=INFO ts=2021-07-22T15:25:29.634753957Z app=edgex-core-metadata source=config.go:193 msg="Loaded configuration from /res/configuration.toml"
level=INFO ts=2021-07-22T15:25:29.636021126Z app=edgex-core-metadata source=variables.go:394 msg="Variables override of 'Registry.Host' by environment variable: REGISTRY_HOST=edgex-core-consul"
level=INFO ts=2021-07-22T15:25:29.63609262Z app=edgex-core-metadata source=variables.go:394 msg="Variables override of 'Service.Host' by environment variable: SERVICE_HOST=edgex-core-metadata"
level=INFO ts=2021-07-22T15:25:29.636189056Z app=edgex-core-metadata source=variables.go:394 msg="Variables override of 'Databases.Primary.Host' by environment variable: DATABASES_PRIMARY_HOST=edgex-redis"
level=INFO ts=2021-07-22T15:25:29.636239947Z app=edgex-core-metadata source=variables.go:394 msg="Variables override of 'Notifications.Sender' by environment variable: NOTIFICATIONS_SENDER=edgex-core-metadata"
level=INFO ts=2021-07-22T15:25:29.636327407Z app=edgex-core-metadata source=variables.go:394 msg="Variables override of 'Clients.CoreData.Host' by environment variable: CLIENTS_COREDATA_HOST=edgex-core-data"
level=INFO ts=2021-07-22T15:25:29.636385095Z app=edgex-core-metadata source=variables.go:394 msg="Variables override of 'Clients.Notifications.Host' by environment variable: CLIENTS_NOTIFICATIONS_HOST=edgex-support-notifications"
level=INFO ts=2021-07-22T15:25:29.636653443Z app=edgex-core-metadata source=config.go:168 msg="Using Configuration provider (consul) from: http://edgex-core-consul:8500 with base path of edgex/core/1.0/edgex-core-metadata"
level=INFO ts=2021-07-22T15:25:35.783911105Z app=edgex-core-metadata source=config.go:304 msg="Configuration has been pushed to into Configuration Provider (6 envVars overrides applied)"
level=ERROR ts=2021-07-22T15:25:35.784027258Z app=edgex-core-metadata source=logger.go:74 msg="logTarget cannot be blank, using stdout only"
level=INFO ts=2021-07-22T15:25:35.78405802Z app=edgex-core-metadata source=registry.go:80 msg="Using Registry (consul) from http://edgex-core-consul:8500"
level=INFO ts=2021-07-22T15:25:35.806415907Z app=edgex-core-metadata source=database.go:147 msg="Database connected"
level=INFO ts=2021-07-22T15:25:35.80888482Z app=edgex-core-metadata source=database.go:112 msg="Database for V2 API connected"
level=INFO ts=2021-07-22T15:25:35.816228234Z app=edgex-core-metadata source=telemetry.go:86 msg="Telemetry starting"
level=INFO ts=2021-07-22T15:25:35.816275058Z app=edgex-core-metadata source=httpserver.go:99 msg="Web server starting (edgex-core-metadata:48081)"
level=INFO ts=2021-07-22T15:25:35.816339316Z app=edgex-core-metadata source=message.go:50 msg="Service dependencies resolved..."
level=INFO ts=2021-07-22T15:25:35.816358311Z app=edgex-core-metadata source=message.go:51 msg="Starting edgex-core-metadata 1.3.1 "
level=INFO ts=2021-07-22T15:25:35.816376791Z app=edgex-core-metadata source=message.go:55 msg="This is the EdgeX Core Metadata Microservice"
level=INFO ts=2021-07-22T15:25:35.816396635Z app=edgex-core-metadata source=message.go:58 msg="Service started in: 6.183851315s"
level=ERROR ts=2021-07-22T15:25:36.075032185Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
level=ERROR ts=2021-07-22T15:25:36.857371575Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
level=ERROR ts=2021-07-22T15:25:37.587933454Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
level=ERROR ts=2021-07-22T15:25:38.532055123Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
level=ERROR ts=2021-07-22T15:25:38.734072744Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
level=ERROR ts=2021-07-22T15:25:38.75834035Z app=edgex-core-metadata source=handler.go:47 msg="no item found for supplied key: edgex-device-rest"
level=ERROR ts=2021-07-22T15:25:38.765760106Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
level=ERROR ts=2021-07-22T15:25:38.774815337Z app=edgex-core-metadata source=handler.go:47 msg="no item found for supplied key: edgex-device-rfid-llrp"
level=ERROR ts=2021-07-22T15:25:39.207524431Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
level=ERROR ts=2021-07-22T15:25:39.211626219Z app=edgex-core-metadata source=handler.go:47 msg="no item found for supplied key: device-virtual"
level=ERROR ts=2021-07-22T15:27:10.560519413Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
level=ERROR ts=2021-07-22T15:27:10.650220089Z app=edgex-core-metadata source=handler.go:47 msg="Item not found"
</code></pre>

## 🌍  Your Environment
**Deployment Environment:**
- Ubuntu 20.04
- EdgeX Hanoi Docker Compose 
    - No security, default configuration
    - LLRP Services (custom addition)

**EdgeX Version:**
Hanoi

**Anything else relevant?**

These messages were found during validation of the LLRP device and app services by our validation engineer. Unfortunately we do not have any more in-depth logs to provide at the moment. These error logs do not appear to affect the normal operation of EdgeX and our services.

I also observed and filed a similar error message on the llrp service a month ago: https://github.com/edgexfoundry/device-rfid-llrp-go/issues/40

**Comment by cloudxxx8** [comment]
@ajcasagrande @lenny-intel do you have a plan to upgrade the LLRP Device Service to v2?  The error log was not clear in v2, and we made lots of improvement in v2.

I saw you enable the UpdateLastConnected , so it should be important to your service.
https://github.com/edgexfoundry/device-rfid-llrp-go/blob/a792bd60d44e476743a31f63448cab357051592b/cmd/res/configuration.toml#L50

Could you please try to disable it in Device Service and enable its counterpart in core-data and see whether the error still exists?
https://github.com/edgexfoundry/edgex-go/blob/v1.3.1/cmd/core-data/res/configuration.toml#L2

You can use environment variable override to achieve this.

**Comment by lenny-intel** [comment]
> do you have a plan to upgrade the LLRP Device Service to v2?

@cloudxxx8 yes, after releasing it for V1.

**Comment by lenny-intel** [comment]
> I saw you enable the UpdateLastConnected , so it should be important to your service.
@ajcasagrande , is UpdateLastConnected important to the LLRP service?

**Comment by lenny-intel** [comment]
> You can use environment variable override to achieve this.

@ajcasagrande you can add the following override to Core Data's environment section in the compose file:
```
WRITABLE_DEVICEUPDATELASTCONNECTED: "true"
```

**Comment by iain-anderson** [comment]
FYI I checked core-metadata v2 (using the C sdk) and it looks okay, but I notice that updating lastConnected triggers the "device updated" callback - not sure whether this is useful or not?

**Comment by cloudxxx8** [comment]
@iain-anderson I agree with you. This callback looks redundant.  I open an issue for this to discuss https://github.com/edgexfoundry/edgex-go/issues/3641

However, I am not sure whether it is useful in the cache for the Device Service implementation.  May we always update the lastConnected in the cache in SDK?  Then, we can cut the callback related to lastConnected update from core-metadata.

**Comment by ajcasagrande** [comment]
@lenny-intel @cloudxxx8 The `UpdateLastConnected` field is not important to the service, and we are not actively using it. I cannot remember the official reason why we enabled it, but I believe it was more or less just something we figured would be nice to have.

I also did what you guys suggested and disabled it in our service and enabled it in `core-data`. That did work, however it seems like our assumptions of the way the field work were incorrect. I believe we thought it was something that was updated frequently and could be used as a way to tell when devices have been offline for a while.

I will disable the field in v1 release, and will re-visit it when it is upgraded to v2.
